src/playlist-stream.js

```diff
@@ -0,0 +1,89 @@
+const Readable = require('readable-stream')
+const sortBy = require('lodash.sortby')
+const {fetch} = require('./playlist')
+
+class PlaylistStream extends Readable {
+  /**
+   * A stream of songs that have been played. This is gaurenteed to contain no
+   * duplicates and will be in chronological order.
+   * @param {Integer} [minPollingInterval=60000] How often to poll for new
+   * songs, in milliseconds. A reasonable value is 60 * 1000, or 60 seconds,
+   * since the average length of a song is about 3 minutes. If no read calls are
+   * made, then this polling interval may be exceeded, therefore it's
+   * technically a minimum.
+   */
+  constructor (minPollingInterval = 60000) {
+    super({objectMode: true})
+    this._minPollingInterval = minPollingInterval
+    this._mostRecentPlayedAt = null
+    this._lastPollTime = null
+    this._pollTimeout = null
+
+    // if the stream is destroyed in the middle of _pollForSongs, then we should
+    // quit pushing data
+    this._isDestroyed = false
+  }
+
+  _pollForSongs () {
+    fetch().then((songs) => {
+      this._pollTimeout = null
+      this._lastPollTime = Date.now()
+
+      // make sure the stream hasn't been destroyed while we were waiting. we
+      // can't write to a destroyed stream
+      if (this._isDestroyed) return
+
+      // filter out all the songs we've seen before, unless this is the first poll
+      if (this._mostRecentPlayedAt !== null) {
+        songs = songs.filter((song) => (
+          song.playedAt > this._mostRecentPlayedAt
+        ))
+      }
+
+      const totalNewSongs = songs.length
+      if (totalNewSongs === 0) {
+        // if we don't write anything, then the read function won't be called,
+        // so schedule another poll
+        this._schedulePoll()
+        return
+      }
+
+      songs = sortBy(songs, 'playedAt')
+      songs.map(
+        this.push.bind(this)
+      )
+      this._mostRecentPlayedAt = songs[totalNewSongs - 1].playedAt
+    })
+  }
+
+  _schedulePoll (delay = this._minPollingInterval) {
+    this._pollTimeout = setTimeout(
```

Comment:
Why not use `setInterval()` here?

Reply:
When you're implementing a stream, you're supposed to [support back-pressure](https://nodejs.org/en/docs/guides/backpressuring-in-streams/), so if the thing that your stream is being piped into stops consuming your output then you stop producing output. That's why I wait for a `_read` call to run `_schedulePoll`.

Also, if the HTTP requests are super slow (> `_minPollingInterval`), I don't want to create a huge backlog of HTTP requests by doing HTTP requests at a fixed interval... I want to make sure I'm waiting *n* seconds between the end of the last HTTP request and the start of the next one, rather than *n* seconds between the start of each HTTP request.

Reply:
Nice, good call 👍